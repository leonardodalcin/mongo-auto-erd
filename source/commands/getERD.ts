import { makeEntity } from '@entity/makeEntity'
import { IEntity } from '@interfaces/IEntity'
import { getDB } from '@mongo/getDB'
import { getDBCollectionNames } from '@mongo/getDBCollectionNames'
import { Spinner } from 'cli-spinner'
import { mapReduceCollectionProperties } from 'source/mongo/mapReduceCollectionProperties'

const spinner = new Spinner()

// spinner.setSpinnerString()

export async function getERD(
  mongoURI: string,
  databaseName: string
): Promise<IEntity[]> {
  spinner.start()
  spinner.setSpinnerTitle('Connecting to database')
  await getDB(mongoURI, databaseName)
  const collectionNames = await getDBCollectionNames()
  const entities = []
  const indexCount = collectionNames.length
  let currentCollectionIndex = 1
  for (const collectionName of collectionNames) {
    spinner.setSpinnerTitle(`Defining entity for collection: ${collectionName}' +
      ' (${currentCollectionIndex}/${indexCount}`)
    entities.push(await makeEntity(collectionName))
    currentCollectionIndex++
  }
  spinner.stop()
  return entities
}
