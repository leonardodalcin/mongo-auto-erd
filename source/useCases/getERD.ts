import { IEntity } from '@interfaces/IEntity'
import { getDB } from '@mongoWrapper/getDB'
import { getDBCollectionNames } from '@mongoWrapper/getDBCollectionNames'
import { getRelationshipTargetCollectionName } from '@mongoWrapper/getRelationshipTargetCollectionName'
import { getRelationshipType } from '@mongoWrapper/getRelationshipType'
import { mapReduceCollectionProperties } from '@mongoWrapper/mapReduceCollectionProperties'
import { Spinner } from 'cli-spinner'

const spinner = new Spinner()
spinner.setSpinnerString(14)

export async function getERD(
  mongoURI: string,
  databaseName: string,
  outdir: string
): Promise<void> {
  await getDB(mongoURI, databaseName)
  spinner.start()
  const collectionNames = await getDBCollectionNames()
  for (const collectionName of collectionNames) {
    const entity: IEntity = {
      collectionName,
      properties: [],
      relationships: []
    }

    const mapReducedProperties = await mapReduceCollectionProperties(
      collectionName
    )
    for (const property of mapReducedProperties) {
      if (property.values.some((value) => value.type === 'objectId')) {
        entity.relationships.push({
          propertyName: property.name,
          sourceCollectionName: collectionName,
          targetCollectionName:
            (await getRelationshipTargetCollectionName(property)) || '',
          type: (await getRelationshipType(property)) || ''
        })
      }
    }
  }
  spinner.stop()
}
