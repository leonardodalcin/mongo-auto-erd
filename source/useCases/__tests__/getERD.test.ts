import { getERD } from '@useCases/getERD'
import 'reflect-metadata'
import { MongoMemoryServer } from 'mongodb-memory-server-core'
import PhotoModel from '@collections/Photo'
import UserModel from '@collections/User'
import UserGroupModel from '@collections/UserGroup'
import * as mongoose from 'mongoose'
jest.setTimeout(5000000)

describe('getERD', () => {
  let mongod: MongoMemoryServer

  beforeAll(async() => {
    mongod = new MongoMemoryServer()
    await mongoose.connect(await mongod.getUri(), {dbName: await mongod.getDbName()})
    const user = await UserModel.create({name: 'Test user name'})
    await PhotoModel.create({title: 'Test photo name', user: user._id})
    await UserGroupModel.create({name: 'Test user group', users: [user._id]})

  })
  afterAll(async() => {
    await mongod.stop()
  })
  it('should get an ERD', async () => {
    await expect(getERD(
      await mongod.getConnectionString(),
      await mongod.getDbName(),
      './file.json'
    )).resolves.toMatchSnapshot()
  })
})
