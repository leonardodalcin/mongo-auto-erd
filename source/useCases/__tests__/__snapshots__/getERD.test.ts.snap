// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getERD should get an ERD 1`] = `
Array [
  Object {
    "collectionName": "usergroups",
    "properties": Array [
      Object {
        "name": "name",
        "types": Array [
          "string",
        ],
      },
    ],
    "relationships": Array [
      Object {
        "propertyName": "__v",
        "sourceCollectionName": "usergroups",
        "targetCollectionName": "",
        "type": "1",
      },
      Object {
        "propertyName": "_id",
        "sourceCollectionName": "usergroups",
        "targetCollectionName": "usergroups",
        "type": "1",
      },
      Object {
        "propertyName": "users",
        "sourceCollectionName": "usergroups",
        "targetCollectionName": "users",
        "type": "1",
      },
    ],
  },
  Object {
    "collectionName": "users",
    "properties": Array [
      Object {
        "name": "name",
        "types": Array [
          "string",
        ],
      },
    ],
    "relationships": Array [
      Object {
        "propertyName": "__v",
        "sourceCollectionName": "users",
        "targetCollectionName": "",
        "type": "1",
      },
      Object {
        "propertyName": "_id",
        "sourceCollectionName": "users",
        "targetCollectionName": "users",
        "type": "1",
      },
    ],
  },
  Object {
    "collectionName": "photos",
    "properties": Array [
      Object {
        "name": "title",
        "types": Array [
          "string",
        ],
      },
    ],
    "relationships": Array [
      Object {
        "propertyName": "__v",
        "sourceCollectionName": "photos",
        "targetCollectionName": "",
        "type": "1",
      },
      Object {
        "propertyName": "_id",
        "sourceCollectionName": "photos",
        "targetCollectionName": "photos",
        "type": "1",
      },
      Object {
        "propertyName": "user",
        "sourceCollectionName": "photos",
        "targetCollectionName": "users",
        "type": "1",
      },
    ],
  },
]
`;
